lib/src/tokenizer.cc

```diff
@@ -0,0 +1,556 @@
+                throw config_exception("Line " + std::to_string(_line_number) + ": JSON does not allow unescaped "
+                                       + string(1, c) + " in quoted strings, use a backslash escape");
+            } else {
+                result += c;
+                original += c;
+            }
+        }
+
+        // maybe switch to triple quoted string
+        if (result.length() == 0) {
+            char third = next_char_raw();
+            if (third == '"') {
+                original += third;
+                append_triple_quoted_string(result, original);
+            } else {
+                put_back(third);
+            }
+        }
+
+        return make_shared<value>(unique_ptr<config_string>(new config_string(_line_origin, result, config_string_type::QUOTED)),
+                                               original);
+    }
+
+    shared_ptr<token> token_iterator::pull_plus_equals() {
+        int c = next_char_raw();
+        if (c != '=') {
+            throw config_exception("'+' not followed by '=', '" + string(1, c) +
+                "' not allowed after '+'");
+        }
+        return tokens::plus_equals_token();
+    }
+
+    shared_ptr<token> token_iterator::pull_substitution() {
+        // The initial '$' has already been consumed
+        shared_ptr<simple_config_origin> origin = _line_origin;
+        char c = next_char_raw();
+        if (c != '{') {
+            throw config_exception("'$' not follwoed by '{', '" + string(1, c) + "' not allowed after '$'");
+        }
+
+        bool optional = false;
+        c = next_char_raw();
+        if (c == '?') {
+            optional = true;
+        } else {
+            put_back(c);
+        }
+
+        whitespace_saver saver;
+        vector<shared_ptr<token>> expression;
+
+        shared_ptr<token> t;
+        do {
+            t = pull_next_token(saver);
+
+            // note that we avoid validating the allowed tokens inside
+            // the substitution here; we even allow nested substitutions
+            // in the tokenizer. The parser sorts it out.
+            if (t == tokens::close_curly_token()) {
+                // we found the end of the substitution
+                break;
+            } else if (t == tokens::end_token()) {
+                throw config_exception("Substitution '${' was not closed with a '}'");
+            } else {
+                shared_ptr<token> whitespace = saver.check(t->get_token_type(), *origin, _line_number);
+                if (whitespace != nullptr) {
+                    expression.push_back(whitespace);
+                }
+                expression.push_back(t);
+            }
+        } while (true);
+
+        return make_shared<substitution>(_line_origin, optional, expression);
+    }
+
+    shared_ptr<token> token_iterator::pull_next_token(whitespace_saver& saver) {
+        char c = next_char_after_whitespace(saver);
+        if (c == -1) {
+            return tokens::end_token();
```

Comment:
This method usually creates a new token, so don't change it to return a const&. That would cause invalid memory use.